primitives/src/merkle_tree/field_based_mht/smt/big_lazy_merkle_tree.rs

```diff
@@ -0,0 +1,987 @@
+                        return Some(retrieved_elem);
+                    },
+                    Err(e) => {
+                        println!("Could not delete node from cache: {}", e);
+                        return None;
+                    }
+                }
+            },
+            Ok(None) => {
+                return None;
+            },
+            Err(e) => {
+                println!("operational problem encountered: {}", e);
+                return None;
+            },
+        }
+    }
+
+    pub fn insert_to_db(&self, idx: usize, data: F) {
+        let elem = to_bytes!(data).unwrap();
+        let index = bincode::serialize(&idx).unwrap();
+        self.database.put(index, elem).unwrap();
+    }
+
+    pub fn get_from_db(&self, idx: usize) -> Option<F>{
+        let index = bincode::serialize(&idx).unwrap();
+        match self.database.get(index) {
+            Ok(Some(value)) => {
+                let retrieved_elem = F::read(value.as_slice()).unwrap();
+                return Some(retrieved_elem);
+            },
+            Ok(None) => {
+                return None;
+            },
+            Err(e) => {
+                println!("operational problem encountered: {}", e);
+                return None;
+            },
+        }
+    }
+
+    pub fn remove_from_db(&self, idx: usize) -> Option<F>{
+        let index = bincode::serialize(&idx).unwrap();
+        match self.database.get(index.clone()) {
+            Ok(Some(value)) => {
+                let retrieved_elem = F::read(value.as_slice()).unwrap();
+                let res = self.database.delete(index.clone());
+                match res {
+                    Ok(_) => {
+                        return Some(retrieved_elem);
+                    },
+                    Err(e) => {
+                        println!("Could not delete leaf from db: {}", e);
+                        return None;
+                    }
+                }
+
+            },
+            Ok(None) => {
+                return None;
+            },
+            Err(e) => {
+                println!("operational problem encountered: {}", e);
+                return None;
+            },
+        }
+    }
+
+    pub fn check_b_plus_caching_level_down(&mut self, coord: Coord) {
+
+        let left_child_idx = coord.idx * T::MERKLE_ARITY;
+        let left_child_height = 0;
+        let left_coord = Coord { height: left_child_height, idx: left_child_idx };
+
+        let right_child_idx = left_child_idx + 1;
+        let right_child_height = 0;
+        let right_coord = Coord { height: right_child_height, idx: right_child_idx };
+
+        if (!self.state.present_node.contains(&left_coord)) | (!self.state.present_node.contains(&right_coord)) {
+            if self.contains_key_in_cache(coord) {
+                LazyBigMerkleTree::remove_node_from_cache(self, coord);
+            }
+        }
+        return;
+    }
+
+    pub fn check_b_plus_caching(&mut self, coord: Coord) {
```

Comment:
Can't we use this function also in BigMerkleTree ?

Reply:
The same as above.